Week3/api.js

```diff
@@ -0,0 +1,218 @@
+const mysql = require('mysql');
+const util = require('util');
+const { handelError } = require('./dbUtils');
+
+const connection = mysql.createConnection({
+  host: 'localhost',
+  user: 'hyfuser',
+  password: 'hyfpassword',
+  database: 'todos_app',
+});
+
+const execQuery = util.promisify(connection.query.bind(connection));
+connection.connect(handelError('connecting...'));
+
+class Api {
+  async getUsersHandler(req, res) {
+    try {
+      const users = await execQuery(`SELECT * FROM todos_app.users`);
+      if (!users.length) return res.status(404).send(`You need to insert some users!`);
+      res.send(users);
+    } catch (err) {
+      console.log(err.message);
+    }
+  }
+
+  async getUserHandler(req, res) {
+    const userId = req.params.id;
+    try {
+      const user = await execQuery(`SELECT * FROM todos_app.users WHERE User_id = ?`, [userId]);
+      if (!user.length) return res.status(404).send('The user with the given Id was not Found!');
+      res.send(user);
+    } catch (err) {
+      console.log(err);
+    }
+  }
+
+  async createUserHandler(req, res) {
+    try {
+      const { Name, User_id, Email } = req.body;
+      const create = `INSERT INTO todos_app.users (User_id, Name, Email) VALUES (?, ?, ?)`;
+      await execQuery(create, [User_id, Name, Email]);
+      res.end('User was successful created!');
+    } catch (err) {
+      console.error(err.message);
+      res.end(err.message);
+    }
+  }
+
+  async deleteUserHandler(req, res) {
+    try {
+      const users = await execQuery(`SELECT * FROM todos_app.users`);
+      const id = parseInt(req.params.id);
+      const user = users.find(user => user.user_id === id);
+      if (!user) return res.status(404).send('The user with the given Id was not Found!');
+      if (users.length === 1) await execQuery(`Delete FROM todos_app.users where user_id = ?`, id);
+      await execQuery(
+        `DELETE users, todolist FROM users INNER JOIN todolist
+      WHERE users.user_id = todolist.user_id and users.user_id = ?`,
+        id,
+      );
+      res.send('User deleted successfully');
+    } catch (err) {
+      console.log(err);
+    }
+  }
+
+  async updateUser(req, res) {
+    try {
+      const users = await execQuery(`SELECT * FROM todos_app.users`);
+      const id = parseInt(req.params.id);
+      const user = users.find(user => user.user_id === id);
+      if (!user) return res.status(404).send('The user with the given Id was not Found!');
+      const { Name, Email } = req.body;
+      await execQuery(`UPDATE users SET Name = ?, Email = ? WHERE user_id = ${id}`, [Name, Email]);
+      res.end('The update was successful!');
+    } catch (err) {
+      res.end(err);
+    }
+  }
+
+  async getTodolistHandler(req, res) {
+    try {
+      const todolist = await execQuery(`SELECT * FROM todos_app.todolist`);
+      if (!todolist.length) return res.status(404).send(`Create a Todo List!`);
```

Comment:
I think it's ok not to have todolist yet ? I would respond 404 if you try to get a specific id but not if you want a list.